cli/cli-client/src/main/java/org/infinispan/cli/patching/PatchTool.java

```diff
@@ -0,0 +1,474 @@
+         }
+      }
+      return infos;
+   }
+
+   private Version getVersion(Path base) throws IOException {
+      // Load the META-INF/infinispan-version.properties from the lib/infinispan-commons jar
+      Path lib = base.resolve("lib");
+      File[] commons = lib.toFile().listFiles((dir, name) -> name.startsWith("infinispan-commons-") && name.endsWith(".jar"));
+      if (commons == null || commons.length != 1) {
+         throw MSG.patchCannotFindCommons(lib);
+      }
+      URI jarUri = URI.create("jar:" + commons[0].toURI().toString());
+      try (FileSystem zipfs = FileSystems.newFileSystem(jarUri, Collections.emptyMap()); InputStream in = Files.newInputStream(zipfs.getPath("META-INF", "infinispan-version.properties"))) {
+         return Version.from(in);
+      }
+   }
+
+
+   private FileSystem getPatchFile(Path patch, boolean create) throws IOException {
+      if (create && patch.toFile().exists()) {
+         throw MSG.patchFileAlreadyExists(patch);
+      }
+      URI jarUri = URI.create("jar:" + patch.toUri().toString());
+      return FileSystems.newFileSystem(jarUri, create ? Collections.singletonMap("create", "true") : Collections.emptyMap());
+   }
+
+   private FileSystem getPatchFile(Path patch) throws IOException {
+      return getPatchFile(patch, false);
+   }
+
+   private void addFileToZip(FileSystem zipfs, Path basePath, ServerFile file) {
+      try {
+         Path target = zipfs.getPath(file.getVersionedPath().toString());
+         out.println(MSG.patchCreateAdd(target));
+         if (target.getParent() != null) {
+            Files.createDirectories(target.getParent());
+         }
+         Files.copy(basePath.resolve(file.getVersionedPath()), target, StandardCopyOption.REPLACE_EXISTING);
+      } catch (IOException e) {
+         throw MSG.patchCreateError(e);
+      }
+   }
+
+   private static boolean compareFileContents(Path path, Path path2) throws IOException {
+      if (Files.isSameFile(path, path2)) {
+         return true;
+      }
+      byte[] buffer1 = new byte[BUFFER_SIZE];
+      byte[] buffer2 = new byte[BUFFER_SIZE];
+      try (InputStream in1 = Files.newInputStream(path);
+           InputStream in2 = Files.newInputStream(path2);) {
+         while (true) {
+            int n1 = in1.read(buffer1);
+            int n2 = in2.read(buffer2);
+            if (n1 != n2) {
+               return false;
+            }
+            if (!Arrays.equals(buffer1, buffer2)) {
+               return false;
+            }
+            if (n1 < BUFFER_SIZE) { // EOF and everything is equals
+               return true;
+            }
+         }
+      }
+   }
+
+   static String sha256(Path path) {
+      try (InputStream in = Files.newInputStream(path)) {
+         MessageDigest digest = MessageDigest.getInstance("SHA-256");
+         byte[] buffer = new byte[BUFFER_SIZE];
+         int read = 0;
+         do {
+            read = in.read(buffer);
+            digest.update(buffer);
```

Comment:
```suggestion
            digest.update(buffer, 0, read);
```

Reply:
I guess you could use the `MapByteBuffer` directly since the files should be small, and do something like this:
```
      try (FileChannel f = (FileChannel) Files.newByteChannel(path, StandardOpenOption.READ)) {
         MappedByteBuffer byteBuffer = f.map(FileChannel.MapMode.READ_ONLY, 0, f.size());
         MessageDigest digest = MessageDigest.getInstance("SHA-256");
         digest.update(byteBuffer);
         return Util.toHexString(digest.digest());
      } catch (IOException | NoSuchAlgorithmException e) {
         e.printStackTrace();
      }

```

Reply:
Nice one